### PR TITLE
[INTERNAL] UI5 Tooling v4 migration guide

### PR DESCRIPTION
**Pull request by d3xter666**
JIRA: CPOUI5FOUNDATION-847

Some of the required documentation would be handled with: https://github.com/SAP/ui5-tooling/pull/963

This migration guide tries to reflects changes described here: https://github.com/SAP/ui5-tooling/issues/701

### CODE REVIEW

**Review by flovogt — CHANGES_REQUESTED** [review]

File: docs/updates/migrate-v4.md [hunk 1]
```diff
@@ -9,59 +9,85 @@
 
 ## Node.js and npm Version Support
 
-To be defined
+**This release requires Node.js versions v20.11.0, v21.2.0 or higher as well as npm v10 or higher.**
+Support for older Node.js and npm releases has been dropped and will cause an error to be shown.
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:13]:
```suggestion
Support for older Node.js and npm releases has been dropped and will cause an error.
```

File: docs/updates/migrate-v4.md [hunk 2]
```diff
@@ -9,59 +9,85 @@
...
 
 ## Specification Versions Support
 
-To be defined
+Going forward, **only projects with Specification Versions 2.0 and higher are supported.**
+
+In case a legacy specification version is detected, **an automatic migration is attempted.**
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:19]:
```suggestion
If a legacy specification version is detected, **an automatic migration is attempted.**
```

File: docs/updates/migrate-v4.md [hunk 3]
```diff
@@ -9,59 +9,85 @@
...
+This means your old projects might still work. Unless they have non-standard configuration in their ui5.yaml.
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:20]:
```suggestion
Your old projects might therefore still work unless they have a non-standard configuration in their ui5.yaml.
```

File: docs/updates/migrate-v4.md [hunk 4]
```diff
@@ -9,59 +9,85 @@
...
 
 ## Changes for Projects
 
-To be defined
+!!! info
+    ✅ Projects defining **Specification Version 3.x** are expected to be **fully compatible with UI5 Tooling v4**
+
+For projects defining the latest **Specification Versions 4.0 and higher**, some changes apply:
+
+* **Breaking Change:** Remove bundle option `usePredefineCalls`. UI5 CLI v4.0.0 and above will always use `sap.ui.predefine` calls in bundles, making this option obsolete. See [Configuration](../pages/Configuration.md#properties) for details.
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:29]:
```suggestion
* **Breaking Change:** Remove the bundle option `usePredefineCalls`. UI5 CLI v4.0.0 and above will always use `sap.ui.predefine` calls in bundles, making this option obsolete. See [Configuration](../pages/Configuration.md#properties) for details.
```

File: docs/updates/migrate-v4.md [hunk 5]
```diff
@@ -9,59 +9,85 @@
...
+
+* **Breaking Change:** New option `async` for `builder.bundles.bundleDefinition.section`. Only applicable if mode = “require”. Default value `true`. See [Configuration: `bundleDefinition.section`](../pages/Configuration.md#properties) for details.
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:31]:
```suggestion
* **Breaking Change:** New `async` option for `builder.bundles.bundleDefinition.section` with default value = `true`; only applicable if mode = “require”. See [Configuration: `bundleDefinition.section`](../pages/Configuration.md#properties) for details.
```

File: docs/updates/migrate-v4.md [hunk 6]
```diff
@@ -9,59 +9,85 @@
...
+
+See also [Configuration: Specification Version 4.0](../pages/Configuration.md#specification-version-40).
+
+## Migrate Your Code 
+
+
+### Changes to @ui5/cli
+
+Occurrences of `console.log` in the codebase are replaced with `process.stderr`.
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:40]:
```suggestion
Occurrences of `console.log` in the code base are replaced with `process.stderr`.
```

File: docs/updates/migrate-v4.md [hunk 7]
```diff
@@ -9,59 +9,85 @@
...
+Аny scripts that capture the output from `console.log` need to be adjusted to rely on `process.stderr`.
+
+### Changes to @ui5/fs
+
+Non-public DuplexCollection#byGlobSource API has been removed.
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:45]:
```suggestion
Non-public `DuplexCollection#byGlobSource` API has been removed.
```

File: docs/updates/migrate-v4.md [hunk 8]
```diff
@@ -9,59 +9,85 @@
...
 
-### Migrate Your Code
+### Changes to @ui5/builder
 
-To be defined
+- **Bundling**: The `usePredefineCalls` option has been removed. Bundling now enforces the use of `sap.ui.predefine` instead of function wrappers.  
 
-## Changes to @ui5/cli
+- **Task API**: The `namespace` option has been renamed to `projectNamespace`. Check documentation for [CustomTasks API](../pages/extensibility/CustomTasks.md#task-implementation)
```
Comment (KlattG) [inline docs/updates/migrate-v4.md:51]:
```suggestion
- **Task API**: The `namespace` option has been renamed to `projectNamespace`. For more information, check the documentation for [CustomTasks API](../pages/extensibility/CustomTasks.md#task-implementation)  
```

**Review by KlattG — CHANGES_REQUESTED** [review]

File: docs/updates/migrate-v4.md [hunk 9]
```diff
@@ -9,59 +9,85 @@
...
+Going forward, **only projects with Specification Versions 2.0 and higher are supported.**
+
+If a legacy specification version is detected, **an automatic migration is attempted.**
+Your old projects might therefore still work unless they have a non-standard configuration in their ui5.yaml.
 
 ## Changes for Projects
 
-To be defined
+!!! info
+    ✅ Projects defining **Specification Version 3.x** are expected to be **fully compatible with UI5 Tooling v4**
```
Comment (RandomByte) [inline docs/updates/migrate-v4.md:25]:
```suggestion
    ✅ Projects defining **Specification Version 2.x** are expected to be **fully compatible with UI5 Tooling v4**
```

There shouldn't be any difference between Specification Version 2 and 3 in this regard

File: docs/updates/migrate-v4.md [hunk 10]
```diff
@@ -9,59 +9,85 @@
...
+
+For projects defining the latest **Specification Versions 4.0 and higher**, some changes apply:
+
+* **Breaking Change:** Remove the bundle option `usePredefineCalls`. UI5 CLI v4.0.0 and above will always use `sap.ui.predefine` calls in bundles, making this option obsolete. See [Configuration](../pages/Configuration.md#properties) for details.
+
+* **Breaking Change:** New `async` option for `builder.bundles.bundleDefinition.section` with default value = `true`; only applicable if mode = “require”. See [Configuration: `bundleDefinition.section`](../pages/Configuration.md#properties) for details.
```
Comment (RandomByte) [inline docs/updates/migrate-v4.md:31]:
```suggestion
* **Breaking Change:** New `async` option for `builder.bundles.bundleDefinition.section` with default value = `true`; only applicable if mode = "require". See [Configuration: `bundleDefinition.section`](../pages/Configuration.md#properties) for details.
```

**Review by RandomByte — CHANGES_REQUESTED** [review]

**Review by RandomByte — APPROVED** [review]

**Pull request merged**